Initial addition of Kafka metrics. (#2485)
open-telemetry/opentelemetry-specification: specification/metrics/semantic_conventions/instrumentation/kafka.md

# Instrumenting Kafka

**Status**: [Experimental](../../../document-status.md)

This document defines how to apply semantic conventions when instrumenting Kafka.

<!-- toc -->

- [Kafka Metrics](#kafka-metrics)
- [Kafka Producer Metrics](#kafka-producer-metrics)

<!-- tocstop -->

## Kafka Metrics

**Description:** General Kafka metrics.

| Name                                 | Instrument    | Value type | Unit   | Unit ([UCUM](../README.md#instrument-units)) | Description    | Attribute Key | Attribute Values |
| ------------------------------------ | ------------- | ---------- | ------ | -------------------------------------------- | -------------- | ------------- | ---------------- |

> | kafka.message.count                  | Counter       | Int64      | messages | `{messages}` | The number of messages received by the broker. | | |

Reviewer: ```suggestion
| kafka.messages                 | Counter       | Int64      | messages | `{messages}` | The number of messages received by the broker. | | |
```

The semantic conventions guidelines suggest pluralizing things that are being counted, not using the "count" or "total" suffix.

Author: I like it.

> | kafka.messages                       | Counter       | Int64      | messages | `{messages}` | The number of messages received by the broker. | | |

Reviewer 2: If we follow https://github.com/open-telemetry/opentelemetry-specification/pull/2525#issuecomment-1119752082, maybe this should be called "eventing.kafka.messages"? (I don't have strong opinion here)

| kafka.leader.elections               | Counter       | Int64      | elections | `{elections}` | Leader election rate (increasing values indicates broker failures). | | |
| kafka.leader.unclean-elections       | Counter       | Int64      | elections | `{elections}` | Unclean leader election rate (increasing values indicates broker failures). | | |

## Kafka Producer Metrics

**Description:** Kafka Producer level metrics.

| Name                                 | Instrument    | Value type | Unit   | Unit ([UCUM](../README.md#instrument-units)) | Description    | Attribute Key | Attribute Values |
| ------------------------------------ | ------------- | ---------- | ------ | -------------------------------------------- | -------------- | ------------- | ---------------- |
| kafka.producer.outgoing-byte.rate    | UpDownCounter | Double     | bytes | `by`| The average number of outgoing bytes sent per second to all servers. | `client-id` | `client-id` value |
| kafka.producer.response.rate         | UpDownCounter | Double     | responses | `{responses}` | The average number of responses received per second. | `client-id` | `client-id` value |

> | kafka.producer.byte-rate             | UpDownCounter | Double     | bytes | `by` | The average number of bytes sent per second for a specific topic. | `client-id` | `client-id` value |

Reviewer: Is this correct?

> | kafka.producer.compression-rate      | UpDownCounter | Double     | compression rate | `{compressionrate}` | The average compression rate of record batches for a specific topic. | `client-id` | `client-id` value |

Reviewer: ```suggestion
| kafka.producer.compression-ratio      | Gauge | Double     | compression ratio | `{compression}` | The average compression rate of record batches for a specific topic. | `client-id` | `client-id` value |
```

|                                      |               |            |                  |                     |                                                                      | `topic`     | topic name        |

> | kafka.producer.record-error-rate     | UpDownCounter | Double     | error rate | `{errorrate}` | The average per-second number of record sends that resulted in errors for a specific topic.  | `client-id` | `client-id` value |

Reviewer: ```suggestion
| kafka.producer.record-errors.rate     | Gauge | Double     | error rate | `{errors}/s` | The average per-second number of record sends that resulted in errors for a specific topic.  | `client-id` | `client-id` value |
```

|                                      |               |            |            |               |                                                                                              | `topic`     | topic name        |

> | kafka.producer.record-retry-rate     | UpDownCounter | Double     | retry rate | `{retryrate}` | The average per-second number of retried record sends for a specific topic. | `client-id` | `client-id` value  |

Reviewer: ```suggestion
| kafka.producer.record-retry.rate     | Gauge | Double     | retry rate | `{retries}/s` | The average per-second number of retried record sends for a specific topic. | `client-id` | `client-id` value  |
```

|                                      |               |            |            |               |                                                                             | `topic`     | topic name         |

> | kafka.producer.record-send-rate      | UpDownCounter | Double     | records sent rate | `{recordssentrate}` | The average number of records sent per second for a specific topic.  | `client-id` | `client-id` value  |

Reviewer: ```suggestion
| kafka.producer.records-sent.rate      | Gauge | Double     | records sent rate | `{records_sent}/s` | The average number of records sent per second for a specific topic.  | `client-id` | `client-id` value  |
```

> | kafka.producer.outgoing-bytes.rate   | Gauge         | Double     | bytes per second | `by`| The average number of outgoing bytes sent per second to all servers. | `client-id` | `client-id` value |
> | kafka.producer.responses.rate        | Gauge         | Double     | responses per second | `{responses}` | The average number of responses received per second. | `client-id` | `client-id` value |
> | kafka.producer.bytes.rate            | Gauge         | Double     | bytes per second | `by` | The average number of bytes sent per second for a specific topic. | `client-id` | `client-id` value |
> |                                      |               |            |       |      |                                                                   | `topic`     | topic name        |
> | kafka.producer.compression-ratio     | Gauge         | Double     | compression ratio | `{compression}` | The average compression ratio of record batches for a specific topic. | `client-id` | `client-id` value |
> |                                      |               |            |                  |                     |                                                                      | `topic`     | topic name        |
> | kafka.producer.record-error.rate     | Gauge         | Double     | error rate | `{error}s` | The average per-second number of record sends that resulted in errors for a specific topic.  | `client-id` | `client-id` value |
> |                                      |               |            |            |               |                                                                                              | `topic`     | topic name        |
> | kafka.producer.record-retry.rate     | Gauge         | Double     | retry rate | `{retries}` | The average per-second number of retried record sends for a specific topic. | `client-id` | `client-id` value  |
> |                                      |               |            |            |               |                                                                             | `topic`     | topic name         |
> | kafka.producer.record-sent.rate      | Gauge         | Double     | records sent rate | `{records_sent}` | The average number of records sent per second for a specific topic.  | `client-id` | `client-id` value  |

Reviewer: ```suggestion
| kafka.producer.outgoing-bytes.rate   | Gauge         | Double     | bytes per second | `by`/s | The average number of outgoing bytes sent per second to all servers. | `client-id` | `client-id` value |
| kafka.producer.responses.rate        | Gauge         | Double     | responses per second | `{responses}/s` | The average number of responses received per second. | `client-id` | `client-id` value |
| kafka.producer.bytes.rate            | Gauge         | Double     | bytes per second | `by/s` | The average number of bytes sent per second for a specific topic. | `client-id` | `client-id` value |
|                                      |               |            |       |      |                                                                   | `topic`     | topic name        |
| kafka.producer.compression-ratio     | Gauge         | Double     | compression ratio | `{compression}` | The average compression ratio of record batches for a specific topic. | `client-id` | `client-id` value |
|                                      |               |            |                  |                     |                                                                      | `topic`     | topic name        |
| kafka.producer.record-error.rate     | Gauge         | Double     | error rate | `{errors}/s` | The average per-second number of record sends that resulted in errors for a specific topic.  | `client-id` | `client-id` value |
|                                      |               |            |            |               |                                                                                              | `topic`     | topic name        |
| kafka.producer.record-retry.rate     | Gauge         | Double     | retry rate | `{retries}/s` | The average per-second number of retried record sends for a specific topic. | `client-id` | `client-id` value  |
|                                      |               |            |            |               |                                                                             | `topic`     | topic name         |
| kafka.producer.record-sent.rate      | Gauge         | Double     | records sent rate | `{records_sent}/s` | The average number of records sent per second for a specific topic.  | `client-id` | `client-id` value  |
```

# Instrumenting Kafka

**Status**: [Experimental](../../../document-status.md)

This document defines how to apply semantic conventions when instrumenting Kafka.

<!-- toc -->

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- tocstop -->

## Kafka Metrics

**Description:** General Kafka metrics.

> | Name                                 | Instrument           | Value type | Unit | Unit ([UCUM](README.md#instrument-units)) | Description    | Attribute Key | Attribute Values |

Author: Done.

| ------------------------------------ | -------------------- | ---------- | ---- | ----------------------------------------- | -------------- | ------------- | ---------------- |
| kafka.message.count                  | Asynchronous Counter | Int64      | messages | `{messages}` | The number of messages received by the broker. | | |
| kafka.request.count                  | Asynchronous Counter | Int64      | requests | `{requests}` | The number of requests received by the broker. | `type` | `produce`, `fetch` |
| kafka.request.failed                 | Asynchronous Counter | Int64      | requests | `{requests}` | The number of requests to the broker resulting in a failure. | `type`  | `produce`, `fetch` |
| kafka.request.time.total             | Asynchronous Counter | Int64      | milliseconds | `ms` | The total time the broker has taken to service requests. | `type` | `Produce`, `FetchConsumer`, `FetchFollower` |
| kafka.request.time.50p               | Asynchronous Gauge   | Double     | milliseconds | `ms` | The 50th percentile time the broker has taken to service requests. | `type` | `Produce`, `FetchConsumer`, `FetchFollower` |

> | kafka.request.time.99p               | Asynchronous Gauge   | Double     | milliseconds | `ms` | The 99th percentile time the broker has taken to service requests. | `type` | `Produce`, `FetchConsumer`, `FetchFollower` |

Reviewer 3: Summary definitely seems like a good backwards-compatible way of representing the items being replaced here.  I also agree that new/future instrumentation should prefer histograms to summaries.  Switching to a histogram from the current metrics would be breaking for users.

> | kafka.request.time.99p               | Asynchronous Gauge   | Double     | milliseconds | `ms` | The 99th percentile time the broker has taken to service requests. | `type` | `Produce`, `FetchConsumer`, `FetchFollower` |

Reviewer: > Switching to a histogram from the current metrics would be breaking for users.

The statement above is not an OpenTelemetry definition, "breaking" is what happens when a vendor cannot meaningfully use some data, but it's a per-vendor definition. There may be a vendor that provides a consistent user experience when Summary and Histogram data points are mixed. At the very least, the Sum, Count, Min, and Max aggregations are well defined in both cases, so vendors have options.

> | kafka.request.queue                  | Asynchronous Gauge   | Int64      | requests | `{requests}` | The number of requests in the request queue. | | |

Reviewer: ```suggestion
| kafka.request.queue                  | UpDownCounter   | Int64      | requests | `{requests}` | The number of requests in the request queue. | | |
```
There's a pattern here, if the description starts with "The number of ..." and the type is Gauge, switch to UpDownCounter. Or, if the type is Gauge and the units are "{pseudo_units}", switch to UpDownCounter.

| kafka.network.io                     | Asynchronous Counter | Int64      | bytes | `by` | The bytes received or sent by the broker. | `state` | `in`, `out` |

> | kafka.purgatory.size                 | Asynchronous Gauge   | Int64      | requests | `{requests}` | The number of requests waiting in the purgatory.  | `type` | `produce`, `fetch` |

Reviewer: ```suggestion
| kafka.purgatory.size                 | UpDownCounter   | Int64      | requests | `{requests}` | The number of requests waiting in the purgatory.  | `type` | `produce`, `fetch` |
```

> | kafka.partition.count                | Asynchronous Gauge   | Int64      | partitions | `{partitions}` | The number of partitions in the broker.  | | |

Reviewer: ```suggestion
| kafka.partition.count                | UpDownCounter   | Int64      | partitions | `{partitions}` | The number of partitions in the broker.  | | |
```

Does this include offline partitions?

> | kafka.partition.offline              | Asynchronous Gauge   | Int64      | partitions | `{partitions}` | The number of offline partitions. | | |

Reviewer: ```suggestion
| kafka.partition.offline              | UpDownCounter   | Int64      | partitions | `{partitions}` | The number of offline partitions. | | |
```

> | kafka.partition.under_replicated     | Asynchronous Gauge   | Int64      | partition  | `{partitions}` | The number of under replicated partitions. | | |

Reviewer: ```suggestion
| kafka.partition.under_replicated     | UpDownCounter   | Int64      | partition  | `{partitions}` | The number of under replicated partitions. | | |
```

> | kafka.isr.operation.count            | Asynchronous Counter | Int64      | operations | `{operations}` | The number of in-sync replica shrink and expand operations. | `operation` | `shrink`, `expand` |

Reviewer: ```suggestion
| kafka.isr.operation.count            | UpDownCounter | Int64      | operations | `{operations}` | The number of in-sync replica shrink and expand operations. | `operation` | `shrink`, `expand` |
```

Author: Any reason to change this to an UpDownCounter? Other instrumentations report this as a rate/counter.

> | kafka.max.lag                        | Asynchronous Gauge   | Int64      | messages   | `{messages}`   | Max lag in messages between follower and leader replicas. | | |

Reviewer: I wonder if there's a more-primitive metric, such as `kafka.follower.lag`, an UpDownCounter, that reports the number of messages each replica is lagging. Then the user can query this for the maximum value.

(Otherwise I agree this should be Gauge, not UpDownCounter, because you would not want to sum maximum values, the only reasonable aggregation of these is to compute the MAX over a group.)

Author: So actually I found this:

MaxLag (used for this metric): "Max lag in messages btw follower and leader replicas".
ConsumerLag (not used at all here): "Lag in messages per follower replica" (additional information includes topic/partition ids).

Sounds like the latter indeed has what you asked. Maybe then remove `kafka.max.lag` and go with `kafka.follower.lag` instead?

> | kafka.controller.active.count        | Asynchronous Gauge   | Int64      | controllers | `{controllers}` | Number of active controllers in the broker. | | |

Reviewer: ```suggestion
| kafka.controller.active.count        | UpDownCounter   | Int64      | controllers | `{controllers}` | Number of active controllers in the broker. | | |
```

> | kafka.leader.election.rate           | Asynchronous Counter | Int64      | elections | `{elections}` | Leader election rate (increasing values indicates broker failures). | | |

Reviewer: ```suggestion
| kafka.leader.elections           | Counter | Int64      | elections | `{elections}` | The number of leader elections (increasing values indicates broker failures). | | |
```

 I don't like attaching "rate" to something that is a count because the rate of change is a meaningful signal from every count.

Author: So the pattern here (as seen by what JMX reports) is that anything that is a **count-per-second** receives a `rate` in the name. With that clarification, should we still strip the `rate` part?

Reviewer: My understanding is that metrics systems like Prometheus and data formats / APIs like OpenTelemetry encourage counting events instead of precomputing rates, because counter data aggregates correctly and rates can be correctly computed from counters, while aggregating rates directly leads to average rates but cannot be used to compute global rates.

That is why I would strip the `.rate` and specify the convention as a Counter. I understand that current instrumentation is outputting the precomputed `.rate` of a counter. I might then write a generic semantic convention that metrics named `X.rate` for `X` with units `U` are semantically equal to be the rate of `X` with units `U/s` as a Gauge series.

This allows dashboards to be built from existing instrumentation; existing instrumentation can be migrated to use Counters; metric processors can be written to convert from `X.rate` to `X` and `X` to `X.rate` as a way to assist with migration, and so on.

> | kafka.unclean.election.rate          | Asynchronous Counter | Int64      | elections | `{elections}` | Unclean leader election rate (increasing values indicates broker failures). | | |

Reviewer: What defines an unclean leader election? Does the total on line 34 include the unclean values on line 35?

> | kafka.unclean.election.rate          | Asynchronous Counter | Int64      | elections | `{elections}` | Unclean leader election rate (increasing values indicates broker failures). | | |

Author: Actually I haven't been able to confirm one or the another, as the docs do not throw enough light into this - however, my impression, based on running tests, is that they are proportional, i.e. `election.rate` includes the `unclean-election.rate` values. WTDY?

> | kafka.logs.flush.time.count          | Asynchronous Counter | Int64      | milliseconds | `ms` | The time it has taken to flush logs. | | |
> | kafka.logs.flush.time.50th           | Asynchronous Gauge   | Double     | milliseconds | `ms` | The 50th percentile time it has taken to flush logs. | | |
> | kafka.logs.flush.time.99th           | Asynchronous Gauge   | Double     | milliseconds | `ms` | The 99th percentile time it has taken to flush logs. | | |

Reviewer: ```suggestion
| kafka.logs.flush.duration         | Histogram | Int64      | milliseconds | `ms` | The timing of logs flushing. | | |
```

Same remarks as for `kafka.request.duration`. Semantically a histogram, acceptable to downgrade to Summary.

## Kafka Producer Metrics

**Description:** Kafka Producer level metrics.

| Name                                 | Instrument         | Value type | Unit | Unit ([UCUM](README.md#instrument-units)) | Description    | Attribute Key | Attribute Values |
| ------------------------------------ | ------------------ | ---------- | ---- | ----------------------------------------- | -------------- | ------------- | ---------------- |

> | kafka.producer.io-wait-time-ns-avg   | Asynchronous Gauge | Double     | nanoseconds | `ns` | The average length of time the I/O thread spent waiting for a socket ready for reads or writes. | client-id | `client-id` value |

Reviewer: ```suggestion
| kafka.producer.io-wait.duration   | Histogram | Double     | nanoseconds | `ns` | The length of time the I/O thread spent waiting for a socket ready for reads or writes. | client-id | `client-id` value |
```

A zero bucket histogram defines Sum and Count which contain the average value.   This additionally allows plotting the number of IO waits as distinct from the average value.

I mean to say that semantically this is a Histogram instrument, but then in a standard configuration for Kafka it would be appropriate to configure a View that selects a zero bucket histogram to lower cost.

> | kafka.producer.outgoing-byte-rate    | Asynchronous Gauge | Double     | bytes | `by`| The average number of outgoing bytes sent per second to all servers. | `client-id` | `client-id` value |

Reviewer: ```suggestion
| kafka.producer.server.output  | Counter | Int64     | `{bytes}` |  The number of output bytes sent to each server. | `client-id` | `client-id` value |
```

Rates should be derived from Counters, generally, I think. Note OTel has a convention for bi-direcitonal transfer counters that end in `.io` but this is uni-directional.

> | kafka.producer.request-latency-avg   | Asynchronous Gauge | Double     | milliseconds | `ms` | The average request latency. | `client-id` | `client-id` value |
> | kafka.producer.request-rate          | Asynchronous Gauge | Double     | requests | `{requests}` | The average number of requests sent per second. | `client-id` | `client-id` value |

Reviewer: ```suggestion
| kafka.producer.request.duration   | Histogram | Double     | milliseconds | `ms` | The request latency. | `client-id` | `client-id` value |
```
As above, these are semantically a histogram that can be aggregated somewhere, including in a View configuration, down to the average value and count.

> | kafka.producer.response-rate         | Asynchronous Gauge | Double     | responses | `{responses}` | Responses received per second. | `client-id` | `client-id` value |

Reviewer: ```suggestion
| kafka.producer.responses         | Counter | Int64     | responses | `{responses}` | Responses received. | `client-id` | `client-id` value |
```

> | kafka.producer.byte-rate             | Asynchronous Gauge | Double     | bytes | `by` | The average number of bytes sent per second for a specific topic. | `client-id` | `client-id` value |

Reviewer: ```suggestion
| kafka.producer.topic.output             | Asynchronous Gauge | Double     | bytes | `by` | The number of bytes sent for a specific topic. | `client-id`, `topic` | `client-id` value and `topic` name |
```

Following the pattern for `kafka.producer.server.output` suggested above.

|                                      |                    |            |       |      |                                                                   | `topic`     | topic name        |

> | kafka.producer.compression-rate      | Asynchronous Gauge | Double     | compression rate | `{compressionrate}` | The average compression rate of record batches for a specific topic. | `client-id` | `client-id` value |

Reviewer: ```suggestion
| kafka.producer.compression-ratio      | Asynchronous Gauge | Double     | compression rate | _unitless_ | The average compression rate of record batches for a specific topic. | `client-id`, `topic` | `client-id` value and `topic` name|
```

| kafka.producer.record-error-rate     | Asynchronous Gauge | Double     | error rate | `{errorrate}` | The average per-second number of record sends that resulted in errors for a specific topic.  | `client-id` | `client-id` value |
|                                      |                    |            |            |               |                                                                                                | `topic`     | topic name        |
| kafka.producer.record-retry-rate     | Asynchronous Gauge | Double     | retry rate | `{retryrate}` | The average per-second number of retried record sends for a specific topic. | `client-id` | `client-id` value  |
|                                      |                    |            |            |               |                                                                             | `topic`     | topic name         |
| kafka.producer.record-send-rate      | Asynchronous Gauge | Double     | records sent rate | `{recordssentrate}` | The average number of records sent per second for a specific topic.  | `client-id` | `client-id` value  |

> |                                      |                    |            |                   |                     |                                                                      | `topic`     | topic name         |

Reviewer: I suggest continuing to replace Gauges w/ the current average rate by Counters from which average rates can be derived. These should be errors, retries, and records sent by topic.
